service/src/event/ordering_task.rs

```diff
@@ -175,188 +122,310 @@ impl OrderingTask {
+        // If our prev is the init event or already been delivered, we can mark ourselves as deliverable.
+        // If our prev wasn't deliverable yet, we track it and repeat (i.e. follow its prev if we don't have it)
+
+        let mut deliverable_queue = VecDeque::new();
+        let mut undelivered =
+            VecDeque::from_iter(self.cid_map.iter().filter_map(|(cid, ev)| match ev {
+                StreamEvent::Undelivered(meta) => {
+                    assert_eq!(meta.cid, *cid);
+                    Some((meta.cid, meta.prev))
+                }
+                _ => None,
+            }));
+
+        while let Some((cid, prev)) = undelivered.pop_front() {
+            if let Some(prev_event) = self.cid_map.get(&prev) {
+                match prev_event {
+                    StreamEvent::InitEvent(_) | StreamEvent::KnownDeliverable(_) => {
+                        trace!(
+                            %prev,
+                            %cid,
+                            "Found event whose prev is already in memory and IS deliverable!"
+                        );
+                        deliverable_queue.push_back(cid)
+                    }
+                    StreamEvent::Undelivered(_) => {
+                        trace!(
+                            %prev,
+                            %cid,
+                            "Found event whose prev is already in memory but NOT deliverable."
+                        );
+                        // nothing to do until it arrives on the channel
+                    }
+                }
+
+                continue;
+            }
+
+            let prev_event = StreamEvent::load_by_cid(pool, prev).await?;
+            if let Some(known_prev) = prev_event {
+                match &known_prev {
+                    StreamEvent::InitEvent(_) | StreamEvent::KnownDeliverable(_) => {
+                        deliverable_queue.push_back(cid);
+                    }
+                    StreamEvent::Undelivered(undelivered_ev) => {
+                        // we'll try to follow this back to something deliverable
+                        undelivered.push_back((undelivered_ev.cid, undelivered_ev.prev));
+                    }
+                }
+                self.add_event(known_prev);
+            } else {
+                trace!("Found event that depends on another event we haven't discovered yet");
+            }
+        }
+
+        let mut newly_ready = deliverable_queue.clone();
+        while let Some(cid) = newly_ready.pop_front() {
+            if let Some(now_ready_ev) = self.remove_by_prev_cid(&cid) {
+                if let Some(ev) = self.cid_map.get(&now_ready_ev) {
+                    match ev {
+                        StreamEvent::InitEvent(_) | StreamEvent::KnownDeliverable(_) => {
+                            warn!(?ev, "should not have found a deliverable event when we expected only undelivered events!");
```

Comment:
same comment as above on if this should be a fatal error since it indicates programmer error.

Reply:
I'm starting to walk this back though. I do think it's worth marking all init events as fatal, because that should never happen... i'm trying to figure out if there's any possibility of changes in the database and our sets getting out of date but all the mutations should be cancel safe, so I don't think so, but crashing the task would be a huge bummer, even though i'm generally a fan of "asserts in production".